core/threaded/scheduler_GEDF_NP_CI.c

```diff
@@ -0,0 +1,547 @@
+ * scheduler to distribute work. Otherwise, it will wait on
+ * '_lf_sched_instance->_lf_sched_semaphore'.
+ *
+ * @param worker_number The worker number of the worker thread asking for work
+ * to be assigned to it.
+ */
+void _lf_sched_wait_for_work(size_t worker_number) {
+    // First, empty the 'output_reactions' for this worker thread into the
+    // '_lf_sched_instance->_lf_sched_triggered_reactions'.
+    _lf_sched_update_triggered_reactions(worker_number);
+
+    // Second, increment the number of idle workers by 1 and
+    // check if this is the last worker thread to become idle.
+    if (lf_atomic_add_fetch(&_lf_sched_instance->_lf_sched_number_of_idle_workers,
+                            1) ==
+        _lf_sched_instance->_lf_sched_number_of_workers) {
+        // Last thread to go idle
+        DEBUG_PRINT("Scheduler: Worker %d is the last idle thread.",
+                    worker_number);
+        // Call on the scheduler to distribute work or advance tag.
+        if (_lf_sched_try_advance_tag_and_distribute()) {
+            // It's time to stop
+            _lf_sched_signal_stop();
+        }
+    } else {
+        // Not the last thread to become idle.
+        // Wait for work to be released.
+        lf_semaphore_acquire(_lf_sched_instance->_lf_sched_semaphore);
+    }
+}
+
+///////////////////// Scheduler Init and Destroy API /////////////////////////
+/**
+ * @brief Initialize the scheduler.
+ *
+ * This has to be called before other functions of the scheduler can be used.
+ * If the scheduler is already initialized, this will be a no-op.
+ *
+ * @param number_of_workers Indicate how many workers this scheduler will be
+ *  managing.
+ * @param option Pointer to a `sched_params_t` struct containing additional
+ *  scheduler parameters.
+ */
+void lf_sched_init(
+    size_t number_of_workers, 
+    sched_params_t* params
+) {
+    DEBUG_PRINT("Scheduler: Initializing with %d workers", number_of_workers);
+    if(!init_sched_instance(&_lf_sched_instance, number_of_workers, params)) {
+        // Already initialized
+        return;
+    }
+
+    size_t queue_size = INITIAL_REACT_QUEUE_SIZE;
+    if (params != NULL) {
+        if (params->max_reactions_per_level != NULL) {
+            // Recalculate the queue size
+            queue_size = 0;
+            for (size_t i = 0; i <= _lf_sched_instance->max_reaction_level; i++) {
+                queue_size += params->max_reactions_per_level[i];
+            }
+        }
+    }
+    DEBUG_PRINT("Scheduler: Adopting a queue size of %u.", queue_size);
+
+    _lf_sched_instance->_lf_sched_array_of_mutexes =
+        (lf_mutex_t*)calloc(1, sizeof(lf_mutex_t));
+    lf_mutex_init(&_lf_sched_instance->_lf_sched_array_of_mutexes[0]);
+
+    // Reaction queue ordered first by deadline, then by level.
+    // The index of the reaction holds the deadline in the 48 most significant
+    // bits, the level in the 16 least significant bits.
```

Comment:
I know this was inherited, but I'm a bit confused about this now. Shouldn't the deadline go into the least significant bits? @edwardalee?

Reply:
I think this relies on the fact that the deadline is inherited by all upstream reactions. It means that an entire chain of reactions with deadlines can execute before a parallel chain without deadlines.